PersistIQ documentation (#1045)
rudderlabs/rudderstack-docs: docs/destinations/streaming-destinations/persistiq.mdx

| `context.traits.twitterUrl` <br /> `traits.twitterUrl` | `data.twitterUrl` | URL |
| `context.traits.linkedinUrl` <br /> `traits.linkedinUrl` | `data.linkedinUrl` | URL |
| `traits.firstName` <br /> `traits.firstname` <br />`traits.first_name` <br /> `context.traits.firstName` <br />`context.traits.firstname` <br /> `context.traits.first_name` |  `data.first_name` | String  |
| `traits.lastName` <br /> `traits.lastname` <br />`traits.last_name` <br /> `context.traits.lastName` <br />`context.traits.lastname` <br /> `context.traits.last_name` |  `data.last_name` | String  |
| `traits.{x}` <br /> `context.traits.{x}`  |  `data.{x}` | -  |

<div class="infoBlock">
The lead status passed in the <code class="inline-code">traits.status</code>/<code class="inline-code">context.traits.status</code> must already be configured in your PersistIQ dashboard.
</div>

## Group

You can use the <Link to="/event-spec/standard-events/group/">`group`</Link> events to add or remove a lead from a group in PersistIQ. RudderStack uses the `/campaigns` endpoint to send this information to PersistIQ.

A sample `group` call is shown below:

```javascript
rudderanalytics.group("group01", {

>   mailbox_id: "mbid123"

Author: ```suggestion
  mailbox_id: "mbid123",
  operation: "add" // set to remove to remove user from a group
```

}, {
  "externalId": [{
    "type": "persistIqLeadId",
    "id": "abc12345"

>     "operation": "add"

Author: ```suggestion
```

| `context.traits.company` <br />`traits.company` | `lead[0].company_name` | String |
| `context.traits.phone` <br />`traits.phone` | `lead[0].phone` | Phone number |
| `context.traits.facebookUrl` <br />`traits.facebookUrl` | `lead[0].facebookUrl` | URL |
| `context.traits.twitterUrl` <br />`traits.twitterUrl` | `lead[0].twitterUrl` | URL |
| `context.traits.linkedinUrl` <br />`traits.linkedinUrl` | `lead[0].linkedinUrl` | URL |
| `traits.firstName` <br /> `traits.firstname` <br />`traits.first_name` <br /> `context.traits.firstName` <br />`context.traits.firstname` <br /> `context.traits.first_name` |  `lead[0].first_name` | String  |
| `traits.lastName` <br /> `traits.lastname` <br />`traits.last_name` <br /> `context.traits.lastName` <br />`context.traits.lastname` <br /> `context.traits.last_name` |  `lead[0].last_name` | String  |
| `traits.{x}` <br /> `context.traits.{x}`  |  `lead[0].{x}` | -  |

<div class="warningBlock">
RudderStack does not batch events before sending them to PersistIQ. Hence, the index inside the <code class="inline-code">lead</code> array is always <code class="inline-code">0</code>.
</div>

> <div class="infoBlock">

Author: ```suggestion
```

> The lead status passed in the <code class="inline-code">traits.status</code>/<code class="inline-code">context.traits.status</code> must already be configured in your PersistIQ dashboard.

Author: ```suggestion
```

> </div>

Author: ```suggestion
```

### Updating an existing lead

RudderStack automatically updates a lead if the `identify` call contains the `persistIqLeadId` field in the `externalId` object. Otherwise, it creates a new lead.

A sample `identify` call that updates an existing PersistIQ lead is shown below:

```javascript
rudderanalytics.identify("1hKOmRA4GRlm", {
  name: "Alex Keener",
  address: "99 Blue Gum Street",
  email: "alex@example.com"
}, {
  "externalId": [{
    "type": "persistIqLeadId",
    "id": "abc12345"
  }]
})
```

<div class="infoBlock">
For more information on getting your PersistIQ lead ID, refer to the <Link to="#faq">FAQ</Link> section below.
</div>

#### Supported mappings

The following table list the mappings between the RudderStack properties and PersistIQ attributes:

| RudderStack property | PersistIQ property | Data type |
| :--- | :--- | :--- |
| `externalId.persistIqLeadId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span> | `leadId` | String |

> | `traits.status` <br /> `context.traits.status`  |  `status` | String  |

Reviewer: just checking if this property mapping should be there in the **Creating a new lead** section as well.. As we mention the "lead status must already be configured in your PersistIQ dashboard..." note there

Author: thanks - good catch - removed the note from **Creating a new lead**. Had discussed with @anantjain45823 and removed the mapping but missed out on removing the note.

| `context.traits.email` <br /> `traits.email`  | `data.email` | String  |
| `context.traits.country` <br /> `traits.country`  | `data.country` | String |

> | `context.traits.city` `traits.city` | `data.city` | String |

Reviewer: ```suggestion
| `context.traits.city` <br/>`traits.city` | `data.city` | String |
```

similarly in the below rows as well

> | `context.traits.gender` `traits.gender` | `data.gender` | String |

Author: ```suggestion
| `context.traits.gender` <br/> `traits.gender` | `data.gender` | String |
```

> | `context.traits.company` `traits.company` | `data.company_name` | String |

Author: ```suggestion
| `context.traits.company` <br/> `traits.company` | `data.company_name` | String |
```

> | `context.traits.phone` `traits.phone` | `data.phone` | Phone number |

Reviewer: @anantjain45823 Should the data type be **String/Text** instead of **Phone number**?

Author: Let's keep it that. I couldn't recollect the data type we'd finalized for phone numbers.

Reviewer 2: Yeah mainly it just to specify the phone no. Format in string type. But anyway string type would be good

> | `context.traits.phone` `traits.phone` | `data.phone` | Phone number |

Author: ```suggestion
| `context.traits.phone` <br/> `traits.phone` | `data.phone` | Phone number |
```

> | `context.traits.facebookUrl` `traits.facebookUrl` | `data.facebookUrl` | URL |

Author: ```suggestion
| `context.traits.facebookUrl` <br /> `traits.facebookUrl` | `data.facebookUrl` | URL |
```

> | `context.traits.twitterUrl` `traits.twitterUrl` | `data.twitterUrl` | URL |

Author: ```suggestion
| `context.traits.twitterUrl` <br /> `traits.twitterUrl` | `data.twitterUrl` | URL |
```

> | `context.traits.linkedinUrl` `traits.linkedinUrl` | `data.linkedinUrl` | URL |

Author: ```suggestion
| `context.traits.linkedinUrl` <br /> `traits.linkedinUrl` | `data.linkedinUrl` | URL |
```

| `traits.firstName` <br /> `traits.firstname` <br />`traits.first_name` <br /> `context.traits.firstName` <br />`context.traits.firstname` <br /> `context.traits.first_name` |  `data.first_name` | String  |
| `traits.lastName` <br /> `traits.lastname` <br />`traits.last_name` <br /> `context.traits.lastName` <br />`context.traits.lastname` <br /> `context.traits.last_name` |  `data.last_name` | String  |
| `traits.{x}` <br /> `context.traits.{x}`  |  `data.{x}` | -  |

<div class="infoBlock">
The lead status passed in the <code class="inline-code">traits.status</code>/<code class="inline-code">context.traits.status</code> must already be configured in your PersistIQ dashboard.
</div>

## Group

You can use the <Link to="/event-spec/standard-events/group/">`group`</Link> events to add or remove a lead from a group in PersistIQ. RudderStack uses the `/campaigns` endpoint to send this information to PersistIQ.

A sample `group` call is shown below:

```javascript
rudderanalytics.group("group01", {
  mailbox_id: "mbid123"
}, {
  "externalId": [{
    "type": "persistIqLeadId",
    "id": "abc12345"
    "operation": "add"
  }]
})
```

<div class="infoBlock">

> If not explicitly specified, RudderStack automatically sets the <code class="inline-code">traits.operation</code> parameter to <code class="inline-code">add</code>. To remove a user from the group, set this parameter to <code class="inline-code">remove</code>.

Reviewer: ```suggestion
If not explicitly specified, RudderStack automatically sets the <code class="inline-code">externalId.operation</code> parameter to <code class="inline-code">add</code>. To remove a user from the group, set this parameter to <code class="inline-code">remove</code>.
```
need confirmation

Author: @anantjain45823

Reviewer 2: It is inside traits not under externalId

</div>

### Supported mappings

The following table list the mappings between the RudderStack properties and PersistIQ attributes:

| RudderStack property | PersistIQ property | 
| :--- | :--- | 
| `groupId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span> | `campaign_id` |
| `externalId.persistIqLeadId` <br/> <span style="color: #4D4DFF;font-size:12px;">Required</span> | `leadId` | 
| `traits.mailbox_id` <br/>  `context.traits.mailbox_id` | `mailbox_id` |

## FAQ

### Where can I find the PersistIQ API key?

To get your PersistIQ API key, follow these steps:

1. Log in to your [PersistIQ dashboard](https://persistiq.com/app/). 
2. From the bottom left corner, click your profile button and go to **Settings and Billing** > **Integrations**. Your API key will be listed here, as shown:

<img src="../../assets/event-stream-destinations/persistiq-api-key.png" alt="PersistIQ API Key"/>

### How do I get the PersistIQ lead ID and campaign ID?

> To get a prospect's lead ID in the [PersistIQ dashboard](https://persistiq.com/app/), click on it. You can find the ID in the resulting URL, as shown:

Reviewer: ```suggestion
To get a prospect's lead ID in the [PersistIQ dashboard](https://persistiq.com/app/), click on the specific prospect. You can find the ID in the resulting URL, as shown:
```
